Add a mechanism to remind users to rotate personal auth tokens

michalkleiner opened the pull request:
### Description:

The PR adds a scheduled task, new config and a mechanism to send notification email to users when their tokens haven't been rotated over the configured number of days.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Add initial code for a mechanism to remind users to rotate personal auth tokens

michalkleiner committed:
Fix CS

caddoo commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenNotification.php:
```diff
@@ -0,0 +1,78 @@
+        );
+        $email->safeSend();
+
+        if (is_callable($this->onNotificationSent)) {
```
Probably can't be false right? Especially as it's a required property in constructor

caddoo commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenNotification.php:
```diff
@@ -0,0 +1,78 @@
+    public function __construct(
+        string $tokenId,
+        string $tokenName,
+        string $tokenCreationDate,
```
Would it better this being a DateTime object?

sgiehl replied:
We don't use `DateTime` objects much in the code. If something else than a string should be used, then a `Piwik\Date` object.

michalkleiner replied:
Left as a string for now

caddoo commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenNotification.php:
```diff
@@ -0,0 +1,78 @@
+use Piwik\Container\StaticContainer;
+use Piwik\Plugins\UsersManager\Emails\AuthTokenNotificationEmail;
+
+final class AuthTokenNotification
```
Responsibility feels quite big, feels like there are two parts.

A value object so the actual Notification data

And the mechanism to send which would just be `sendNotification` which would take an instance of the value object.

michalkleiner replied:
Moved the sendNotification to the task, the notification now only provides information about which email class to use.

caddoo commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenNotification.php:
```diff
@@ -0,0 +1,78 @@
+use Piwik\Container\StaticContainer;
+use Piwik\Plugins\UsersManager\Emails\AuthTokenNotificationEmail;
+
+final class AuthTokenNotification
```
Also is AuthToken too specific or should it just be Token?

Applies to all classes.

sgiehl replied:
Authtoken seems fine here, as this is the specific implementation around auth tokens only.

michalkleiner replied:
created an interface and an abstract class that can be then extended by plugins if needed

sgiehl commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenProviderInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+namespace Piwik\Plugins\UsersManager\AuthTokenNotifications;
+
+interface AuthTokenProviderInterface
```
This class should better be called `TokenProviderInterface`, as this should be implemented by other plugins, which might not provide AuthTokens but other tokens.

sgiehl commented on plugins/UsersManager/AuthTokenNotifications/AuthTokenProviderInterface.php:
```diff
@@ -0,0 +1,21 @@
+     *
+     * @return AuthTokenNotification[]
+     */
+    public function getAuthTokensToNotify(): array;
```
```suggestion
    public function getTokensToNotify(): array;
```

sgiehl commented on plugins/UsersManager/AuthTokenProvider.php:
```diff
@@ -0,0 +1,76 @@
+        return Date::factory('today')->subDay($periodDays)->getDateTime();
+    }
+
+    public function setTokenNotified(string $tokenId): void
```
As mentioned above, I would suggest to have the `setTokenNotified` part of the interface and always call it after a token was notified.

michalkleiner replied:
I wanted to avoid the provider knowing about how to set tokens being notified, but now it is within the notification class itself and that doesn't need to know about that either, so I can move it around. At some point it would be good to perhaps have a separate model for tokens as currently they are just a side thing for users and don't have a solid structure of their own.

sgiehl reviewed (changes requested):
Good approach. Left some high level comments for possible improvements/changes.

michalkleiner committed:
Adjust token provider interface, make it more reusable, adjust responsibilities

michalkleiner committed:
Mark class as final

michalkleiner commented:
@sgiehl I've changed the responsibilities a bit, added new interface for the token notification and made it all a bit more reusable and extendable. Please have another look.

michalkleiner committed:
Add migration to create new column and bump core version

sgiehl commented on plugins/UsersManager/Emails/AuthTokenNotificationEmail.php:
```diff
@@ -0,0 +1,118 @@
+                );
+        }
+
+        return implode(', ', $parts);
```
This might actually be a bit inaccurate or might have at least unexpected results.
The result currently depends on the current date. If you e.g. configure 28 days and call that method in February it will return `1 Month`. Calling it in any other month will return `28 Days`

michalkleiner replied:
Removed for now in favour of always displaying number of days. We can iterate on this later as it's not the main part of the functionality.

sgiehl commented on plugins/UsersManager/TokenNotifications/TokenNotificationInterface.php:
```diff
@@ -0,0 +1,23 @@
+
+    public function getEmailClass(): string;
+
+    public function getEmailAddress(): string;
```
I might have tried to implement that a bit more generic. This makes only email notifications possible. In theory it could make sense to send different notifications depending on the token. Also in some cases multiple people might need to be notified. Maybe a simple method `send()` would do the trick. So the notification class would handle sending the notification(s).

michalkleiner replied:
Good point, abstracted to a reusable notification class with a `dispatch` method and build on top of that.

sgiehl commented on plugins/UsersManager/TokenNotifications/TokenNotifierTask.php:
```diff
@@ -0,0 +1,99 @@
+     *
+     * @return array
+     */
+    private function getTokensToNotifyByProvider(): array
+    {
+        $tokensToNotifyByProvider = [];
+        $providers = PluginManager::getInstance()->findComponents(
+            'TokenProvider',
+            TokenProviderInterface::class
+        );
+
+        /** @var TokenProvider $provider */
+        foreach ($providers as $provider) {
+            $tokensToNotifyByProvider[get_class($provider)] = [
+                'tokensToNotify' => $provider->getTokensToNotify(),
+                'onTokenNotified' => $provider->onTokenNotified(),
+            ];
+        }
+
+        return $tokensToNotifyByProvider;
+    }
```
I actually can't see the benefit of iterating over all providers to get the details just in order to iterate over the result later again. I would simply return `$providers` here and do the iteration part where it's needed only. 
That way you could also directly call methods of the provider in the code below without having to use callbacks.

michalkleiner replied:
Yep, that makes more sense. Updated.

michalkleiner committed:
Rework responsibilities, abstract token notification for allow other types other than email, simplify token provider, remove callbacks

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

michalkleiner committed:
Remove unnecessary use statement

mneudert commented on plugins/UsersManager/TokenNotifications/TokenNotifierTask.php:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Send token notifications for each provider
+ */
+class TokenNotifierTask extends Task
```
To have this task fully available we also need an entry in the `UsersManager\Tasks::schedule()` function.

Something like `$this->scheduleTask(new TokenNotifierTask());` should work.

mneudert commented on plugins/UsersManager/TokenNotifications/TokenNotifierTask.php:
```diff
@@ -0,0 +1,73 @@
+            Option::set(self::LAST_RUN_TIME_OPTION_NAME, Date::factory('today')->getTimestamp());
+
+            /** @var TokenProviderInterface $provider */
+            foreach ($this->getTokenProviders() as $provider) {
```
The return documentation of `PluginsManager::findComponents` is wrong, it does not return any instances, only a list of class strings.

mneudert commented on plugins/UsersManager/TokenProvider.php:
```diff
@@ -0,0 +1,74 @@
+            $email = $user['email'];
+
+            $notifications[] = new AuthTokenEmailNotification(
+                $t->idusertokenauth,
```
`$t` is an array :see_no_evil:

michalkleiner replied:
🙈

mneudert commented on plugins/UsersManager/TokenNotifications/TokenProviderInterface.php:
```diff
@@ -0,0 +1,29 @@
+
+interface TokenProviderInterface
+{
+    /**
+     * Provides a list of tokens to be notified, each with their information
+     * that can be used to populate the notification email
+     *
+     * @return TokenNotificationInterface[]
+     */
+    public function getTokensToNotify(): array;
```
With the wording "getTokens" I would expect a pure data container, and not an object that has a "dispatch" functionality :thinking: 

Maybe something like `getPendingTokenNotifications`?

mneudert commented on plugins/UsersManager/TokenNotifications/TokenNotificationInterface.php:
```diff
@@ -0,0 +1,21 @@
+
+    public function getTokenCreationDate(): string;
+
+    public function dispatch(): bool;
```
The `TokenNotifierTask` is currently ignoring the return value.

Would it make sense to at least add a "if dispatch returned fales, do not set notified"? Or can we change this to `void` if we don't intend to act upon it?

michalkleiner committed:
Fix typo

michalkleiner committed:
Rename interfaces, classes and methods to better suit their intended functionality

michalkleiner committed:
Ensure TokenNotifierTask is scheduled

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

michalkleiner committed:
Correctly use array access instead of object access to db row data

michalkleiner commented:
Thanks for your feedback @mneudert, all should be addressed.

michalkleiner committed:
Tweaks from further local testing

michalkleiner committed:
Allow to disable auth token notifications

michalkleiner committed:
Exclude system tokens

michalkleiner committed:
Add auth token notification email tests

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

michalkleiner committed:
Fix CS

michalkleiner committed:
Add log info about number of notifications sent

michalkleiner committed:
Declare test fixture variable

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

michalkleiner committed:
Use strings in Tokens fixture

michalkleiner committed:
Add ts_rotation_notified field to expected token test data

michalkleiner committed:
Update UI test screenshot

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

mneudert commented on core/Updates/5.4.0-b1.php:
```diff
@@ -0,0 +1,39 @@
+    public function getMigrations(Updater $updater)
+    {
+        return [
+            $this->migration->db->addColumn('user_token_auth', 'ts_rotation_notified', 'TIMESTAMP NULL'),
```
```suggestion
            $this->migration->db->addColumn('user_token_auth', 'ts_rotation_notified', 'DATETIME NULL'),
```

We should use the same field type in both the schema and the migration. And imho `DATETIME` to not have the confusion of `TIMESTAMP` being affected by MySQL timezone settings.

mneudert commented on plugins/UsersManager/TokenNotificationProvider.php:
```diff
@@ -0,0 +1,80 @@
+
+        $db = Db::get();
+        $sql = "SELECT * FROM " . Common::prefixTable('user_token_auth')
+            . " WHERE (date_expired is null or date_expired > ?)"
```
Got an email for my local "anonymous default token" that was created when the anonymous user was activated for a site.

I think we need to specifically ignore that one.

Afaik there is no way to rotate it, and the user email is `anonymous@example.org` by default, so the noone will receive it.

michalkleiner replied:
Yeah, I've received it locally as well but thought it'll be an email going to nowhere so could be fine. Updated the SQL to exclude tokens for `anonymous@example.org`.

mneudert commented on plugins/UsersManager/TokenNotificationProvider.php:
```diff
@@ -0,0 +1,80 @@
+
+    public function setTokenNotificationDispatched(string $tokenId): void
+    {
+        $this->userModel->setRotationNotificationWasSentForToken($tokenId, $this->today);
```
This always stores the `ts_rotation_notified` with a `00:00:00` time in the database.

For later auditing it might be good to have the real time available. Makes it easier to find relevant logs without having to check a full day :thinking:

michalkleiner replied:
Good catch, updated to use current time.

mneudert commented on plugins/UsersManager/Emails/AuthTokenNotificationEmail.php:
```diff
@@ -0,0 +1,103 @@
+
+    protected function getManageAuthTokensLink(): string
+    {
+        return Url::getCurrentUrlWithoutQueryString()
```
This method is not behaving nicely in a CLI task.

Got a link to `http://unknown/var/www/html/console?module=UsersManager&action=userSecurity#authtokens` this way :see_no_evil: 

The tracking failure notification email uses `SettingsPiwik::getPiwikUrl()`:

https://github.com/matomo-org/matomo/blob/9a3ef94df61682e7d543f3f8b438c368f6eb71a6/plugins/CoreAdminHome/Emails/TrackingFailuresEmail.php#L97

michalkleiner replied:
Thanks, used the suggested way of creating the link.

michalkleiner committed:
Use DATETIME column type in migration

Co-authored-by: Marc Neudert <marc@innocraft.com>

michalkleiner committed:
Update Manage auth token link URL to behave correctly when task run from CLI

michalkleiner committed:
Exclude anonymous user default token from token notifications

michalkleiner committed:
Store current datetime when token notification sent

michalkleiner committed:
Merge branch '5.x-dev' into dev-18658

mneudert commented on plugins/UsersManager/TokenNotificationProvider.php:
```diff
@@ -44,11 +44,13 @@ public function getTokenNotificationsForDispatch(): array
             . " AND ts_rotation_notified is null"
-            . " AND system_token = 0";
+            . " AND system_token = 0"
+            . " AND u.email != 'anonymous@example.org' AND u.invited_by IS NULL";
```
```suggestion
            . " AND u.login != 'anonymous'";
```

Would suggest matching on the username here.

The username is fixed in various checks across the system. Changing that value in the database "converts" the user to a regular one, changing the email however does (to my knowledge) not change the behaviour of the anonymous login in any way.

The `invited_by` check is probably not something we want to have. The value is not reset if the invitation is accepted, requiring it to not be set would deactivate the notifications for nearly every user in an instance. If you actually intended the check to better match the anonymous user and just missed the brackets (`AND (u.email ... u.invited_by)`), using the username removes the necessity for that condition.

michalkleiner replied:
Yeah, I wasn't sure if the user can change somehow, so added more checks than probably necessary as you mentioned the email above as well.

Changed to just check the username.

michalkleiner committed:
Exclude anonymous user by login

michalkleiner commented:
Thanks for the reviews everyone! I'll leave the branch open until internal QA is done on it.
